design_docs/2023_07_20_dynamic_shader_parameter_amount.md

```diff
@@ -0,0 +1,102 @@
+# Passing variables to the shader
+
+We need to be able to pass an arbitrary number of textures and parameters to the custom shaders. For parameters, we need to be able to pass some "global parameters" -- constants used in all videos, as well as some per-video data.
+
+## Possible solutions
+
+### Textures
+
+Unfortunately, it turns out that binding texture arrays is only possible when either
+
+- Textures have the same resolution
+
+or
+
+- The shader always accepts a constant number of textures
+
+This means we need to always bind a constant number of textures. If user's shader needs less textures than the length of the array, we can just fill the rest of it with references to a 1x1 texture which shouldn't harm performance in any way.
+
+Obviously we also need to pass the amount of textures that are actually present in the array. We should use a push constant for this. Push constants are a very efficient mechanism for sending a very small amount of parameters to the shader.
+
+Bringing it all together, the structure of variables declared in the user's shader could look like this:
+
+```wgsl
+@group(0) @binding(0) textures: binding_array<texture_2d<f32>, 16>;
+var<push_constant> textures_len: u32;
+```
+
+### Shader parameters
+
+Binding const-sized uniform/storage buffers works on every backend. This would work approximately like this in the shaders (ofc the len can be defined in a different way, e.g. as the length of the texture array):
+
+```wgsl
+struct MyList {
+    // you declare an array with a constant, large size
+    per_video_data: array<MyStruct, 64>
+    // and then in some way specify how much of this buffer is filled
+    per_video_data_len: u32
+}
+
+@group(1) @binding(1) var<uniform> per_video_data: MyList
+```
+
+We could also use runtime-sized storage buffers, which don't need a constant size (their size is defined at runtime). Unfortunately, binding runtime-sized storage buffers only works on DX12 and Vulkan. Using it would look like this:
+
+```wgsl
+// the array is dynamically sized. 
+// you can get the length using arrayLength(per_video_data)
+@group(1) @binding(1) var<storage> per_video_data: array<MyStruct>
+```
+
+Using this solution it's also possible to make one storage buffer that has some headers and then the dynamically sized per-video data (a runtime-sized array can be the last element of a struct, like in `C`):
+
+```wgsl
+struct MyStruct {
+    some_param: u32,
+    some_other_param: f32,
+    per_video_data: array<MyVideoData>
+}
+```
+
+This is also (sort of) possible using the first solution with a const limit on the video amount:
+
+```wgsl
+struct MyStruct {
+    some_param: u32,
+    some_other_param: f32,
+    per_video_data: array<MyVideoData, 64>
+}
+```
```

Comment:
What about memory alignment here? Wouldn't it require adding some extra "padding"? https://www.w3.org/TR/WGSL/#memory-layouts

Reply:
depends on the alignment of the `MyVideoData`, but potentially yes. The users will have to remember about this, it's not something that we can do for them.